File: src/main/java/play/socketio/SocketIOEventAck.java
/*
 * Copyright (C) 2017 Lightbend Inc. <https://www.lightbend.com>
 */

package play.socketio;

import akka.util.ByteString;
import play.api.libs.json.JsValue;
import scala.Function1;
import scala.Unit;
import scala.collection.Seq;
import scala.runtime.AbstractFunction1;
import scala.util.Either;

/**
 * A socket.io ack function.
 *
 * This is defined in Java so that it can be a functional interface in Scala 2.11.
 */
public interface SocketIOEventAck {
  void apply(Seq<Either<JsValue, ByteString>> args);

  default <T> Function1<T, Unit> compose(Function1<T, Seq<Either<JsValue, ByteString>>> f) {
    return new AbstractFunction1<T, Unit>() {
      @Override
      public Unit apply(T t) {
        SocketIOEventAck.this.apply(f.apply(t));
        return null;
      }
    };
  }

  /**
   * Convenience to create an ack function from Scala.
   */
  static SocketIOEventAck fromScala(Function1<Seq<Either<JsValue, ByteString>>, Unit> f) {
    return f::apply;
  }
}

File: src/main/scala/play/socketio/SocketIOEventAck.scala
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */
package play.socketio

import org.apache.pekko.util.ByteString
import play.api.libs.json.JsValue

/**
 * A socket.io ack function.
 */
trait SocketIOEventAck {

  def apply(args: Seq[Either[JsValue, ByteString]]): Unit

  def compose[T](f: T => Seq[Either[JsValue, ByteString]]): T => Unit = (args: T) => {
    SocketIOEventAck.this.apply(f(args))
  }
}

object SocketIOEventAck {
  def apply(f: Seq[Either[JsValue, ByteString]] => Unit): SocketIOEventAck = args => f(args)

  @deprecated("Use apply method", since = "1.1.0")
  def fromScala(f: Seq[Either[JsValue, ByteString]] => Unit): SocketIOEventAck = apply(f)
}
